src/category_theory/sigma/basic.lean

```diff
@@ -0,0 +1,253 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.natural_isomorphism
+import category_theory.eq_to_hom
+import data.sigma.basic
+import category_theory.pi.basic
+
+/-!
+# Disjoint union of categories
+
+We define the category structure on a sigma-type (disjoint union) of categories.
+-/
+
+namespace category_theory
+namespace sigma
+
+universes w₁ w₂ w₃ v₁ v₂ u₁ u₂
+
+variables {I : Type w₁} {C : I → Type u₁} [Π i, category.{v₁} (C i)]
+
+/--
+The type of morphisms of a disjoint union of categories: for `X : C i` and `Y : C j`, a morphism
+`(i, X) ⟶ (j, Y)` if `i = j` is just a morphism `X ⟶ Y`, and if `i ≠ j` there are no such morphisms.
+-/
+inductive sigma_hom : (Σ i, C i) → (Σ i, C i) → Type (max w₁ v₁ u₁)
+| mk : Π {i : I} {X Y : C i}, (X ⟶ Y) → sigma_hom ⟨i, X⟩ ⟨i, Y⟩
+
+namespace sigma_hom
+
+/-- The identity morphism on an object. -/
+def id : Π (X : Σ i, C i), sigma_hom X X
+| ⟨i, X⟩ := mk (𝟙 _)
+
+instance (X : Σ i, C i) : inhabited (sigma_hom X X) := ⟨id X⟩
+
+/-- Composition of sigma homomorphisms. -/
+def comp : Π {X Y Z : Σ i, C i}, sigma_hom X Y → sigma_hom Y Z → sigma_hom X Z
+| _ _ _ (mk f) (mk g) := mk (f ≫ g)
+
+instance : category_struct (Σ i, C i) :=
+{ hom := sigma_hom,
+  id := id,
+  comp := λ X Y Z f g, comp f g }
+
+@[simp]
+lemma comp_def (i : I) (X Y Z : C i) (f : X ⟶ Y) (g : Y ⟶ Z) :
+  comp (mk f) (mk g) = mk (f ≫ g) :=
+rfl
+
+lemma assoc : ∀ (X Y Z W : Σ i, C i) (f : X ⟶ Y) (g : Y ⟶ Z) (h : Z ⟶ W), (f ≫ g) ≫ h = f ≫ g ≫ h
+| _ _ _ _ (mk f) (mk g) (mk h) :=
+  begin
+    change mk ((f ≫ g) ≫ h) = mk (f ≫ g ≫ h),
+    rw [category.assoc],
+  end
+
+lemma id_comp : ∀ (X Y : Σ i, C i) (f : X ⟶ Y), 𝟙 X ≫ f = f
+| _ _ (mk f) :=
+  begin
+    change mk (𝟙 _ ≫ f) = mk f,
+    rw [category.id_comp],
```

Comment:
This and `comp_id` look like they might have one-line term mode proofs, but if `@`s are involved then maybe it's better to leave things as they are.

Reply:
iirc I couldn't find a one-line term mode proof, the `rw` is doing a rewrite inside of the `mk`, and the deconstruction to `f` additionally unpacks the composition into a nicer version, so I suspect it might not be possible. I'll try again though, now that I'm reconsidering maybe something sneaky with `congr_arg` could work

Reply:
It worked! Thanks